[schema_upgrader] add C++ upgrader for json based upgrading

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #156761

Differential Revision: [D77459912](https://our.internmc.facebook.com/intern/diff/D77459912)

ydwu4 committed:
[schema_upgrader] add C++ upgrader for json based upgrading

[ghstack-poisoned]

ydwu4 committed:
Update on "[schema_upgrader] add C++ upgrader for json based upgrading"




[ghstack-poisoned]

ydwu4 committed:
Update on "[schema_upgrader] add C++ upgrader for json based upgrading"




[ghstack-poisoned]

ydwu4 committed:
Update on "[schema_upgrader] add C++ upgrader for json based upgrading"




[ghstack-poisoned]

ydwu4 committed:
Update on "[schema_upgrader] add C++ upgrader for json based upgrading"




[ghstack-poisoned]

ydwu4 commented on torch/csrc/export/upgrader.cpp:
```diff
@@ -0,0 +1,221 @@
+
+// NOTE: The following version_0 and version_1 upgraders are for testing
+// purposes only. They demonstrate the upgrader system functionality and are
+// used in test/export/test_upgrader.py.
```
Good idea. I put the test upgraders in a seperate file, add a python binding for it and the python test can dynamically register/deregister it. I don't want to add a legit C++ test, the build system is complicated lol

ydwu4 commented on torch/csrc/export/upgrader.cpp:
```diff
@@ -0,0 +1,221 @@
+
+static void registerUpgrader(
+    int version,
+    const std::vector<std::string>& keypath,
```
Add a separate overload for this purpose. My initial thought is that we don't want to hardcode some deliminator into the key path, which can be troublesome somtimes (e.g. the nn_module_stack deliminator).

ydwu4 committed:
Update on "[schema_upgrader] add C++ upgrader for json based upgrading"




[ghstack-poisoned]

ydwu4 committed:
Update on "[schema_upgrader] add C++ upgrader for json based upgrading"




[ghstack-poisoned]

ydwu4 committed:
Update on "[schema_upgrader] add C++ upgrader for json based upgrading"




[ghstack-poisoned]

angelayi commented on torch/csrc/export/upgrader.cpp:
```diff
@@ -0,0 +1,253 @@
+  }
+
+  // Validate that we reached the target version if requested
+  if (validate_target && current_version != target_version) {
```
Wouldn't we always want to validate this?

ydwu4 replied:
sometimes, it could be easier to use not provide a target_version e.g. in testing lol.

angelayi commented on torch/csrc/export/upgrader.cpp:
```diff
@@ -0,0 +1,253 @@
+}
+
+nlohmann::json upgrade(const nlohmann::json& artifact) {
+  return upgradeImpl(artifact, std::numeric_limits<int>::max(), false);
```
In what case would we want to use max? I feel like we would just always want the existing schema version?

angelayi replied:
hm but are you using this anywhere? all your test cases are specifying the version

angelayi commented on torch/csrc/export/pybind.cpp:
```diff
@@ -15,13 +17,34 @@ void initExportBindings(PyObject* module) {
+
+        // Query the current Python schema version as target
+        py::module_ schema_module =
+            py::module_::import("torch._export.serde.schema");
```
I wonder if we could add the schema version to [`generated_serialization_types.h`](https://fburl.com/code/yijogwgt) so that we can query the schema version directly in the `upgrade` cpp function rather than here. cc @zhxchen17 

can leave as a TODO for now

angelayi replied:
> edit: oh, you mean we expose the schema_version in generated_serialization_types.h as a constant? 

yup

angelayi commented on torch/csrc/export/example_upgraders.cpp:
```diff
@@ -0,0 +1,89 @@
+
+  registerUpgrader(
+      0,
+      "graph_module.graph",
```
sorry, maybe change one of these upgraders to go through the vector path?

ydwu4 replied:
sure..

ydwu4 replied:
see the upgrader down below

ydwu4 committed:
Update on "[schema_upgrader] add C++ upgrader for json based upgrading"




[ghstack-poisoned]

angelayi commented on torch/csrc/export/example_upgraders.cpp:
```diff
@@ -0,0 +1,89 @@
+#include <torch/csrc/export/example_upgraders.h>
```
I still wish this file was in test/ but I guess if the build system is too annoying to figure out it's ok

ydwu4 replied:
lol, it's indeed annoying

angelayi reviewed (approved):
letsgoo

you might want to import to internal before landing, just to make sure none of the cpp changes will break

ydwu4 commented:
@ydwu4 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D77459912).

ydwu4 commented:
@pytorchbot merge

ydwu4 commented:
@pytorchbot revert -m "break linter test, which doesn't show up in the pr" -c nosignal

ydwu4 committed:
Update on "[schema_upgrader] add C++ upgrader for json based upgrading"


Differential Revision: [D77459912](https://our.internmc.facebook.com/intern/diff/D77459912)

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge
